### ISSUE TITLE
Changing FX29K's I2C address in register

### ISSUE THREAD

**Issue by jakefave** [eepj/FX29K_for_Arduino#3]
Hi and thank you for creating the library. 

I'm having difficulty translating the given C# code linked on the datasheet to change the address (for example from 0x28 to 0x36) do you have any suggestions?

Thank you

**Comment by eepj** [comment]
Hello. You can pass the I2C address in the object constructor.

The first parameter to the [constructor](https://github.com/eepj/FX29K_for_Arduino/blob/60adc59c6d229ad6248b5f5aa92595a4511f0253/FX29K_for_Arduino.ino#L3) is the I2C address.
```c++
FX29K::FX29K(uint8_t addr, uint8_t range);
```

If you wish to change the I2C address (say to `0x36`) you can pass it to the constructor as the first parameter.
```c++
FX29K scale(0x36, 0010);
```
Or simply pass the load cell's model number. The corresponding address are defined [here](https://github.com/eepj/FX29K_for_Arduino/blob/60adc59c6d229ad6248b5f5aa92595a4511f0253/FX29K.h#L6-L10).
```c++
FX29K scale(FX29K1, 0010);
```

**Comment by eepj** [comment]
I think I might have understood your issue. Do you mean changing the hardware I2C address of the load cell?

I don't know if there's a way to change the I2C address of the load cell. But if you need this because you want to use multiple load cells in parallel you can use a I2C multiplexer like TCA9584.

**Comment by jakefave** [comment]
Yes, I'm trying to change the I2C address of the load cell, or 'slave address' contained within the sensor's EEPROM. Thank you, that is exactly what we did to prototype functionality (use a multiplexer), now I'm trying to simplify the schematics and BOM. On page 14 of the datasheet, TE links C code to accomplish this. However, currently I am not skilled enough to have full confidence I'm doing it properly (having trouble reading where 0x28 is stored in the EEPROM).  Continuing to learn more but I am on limited time (capstone project).

If possible, could you please take a look at what they link to page 14 in the datasheet and relay your thoughts? I've programmed in c++, but this seems like a different beast. Thanks again!

**Comment by eepj** [comment]
Hi. Thanks for clarifying.

I've written a (untested) Arduino sketch to alter the I2C address of a FX29K load cell. Unfortunately I haven't tested the sketch yet as I don't have the load cell at the moment. I'll have to find one from our lab later this week.

Forgive me for having to say this first: by running the sketch you will alter the sensor's internal registers. This will **void your sensor's warranty and may cause it to malfunction**. I **cannot** guarantee that your sensor will still function properly after running the sketch.

But if you really want to try the sketch now you can find it [here](https://github.com/eepj/FX29K_for_Arduino/blob/test/i2c_address_change/examples/ChangeI2CAddress/ChangeI2CAddress.ino).

Cheers!

**Comment by eepj** [comment]
I'll try to elaborate TE's C code and please let me know if you spot any error:

To access the internal registers, we issue I2C commands like how we normally read the weight data. The first byte is the 7-bit old I2C address of the load cell and 1 read/write bit. Usually this is handled by the MCU's I2C library.

The tailing bytes 2-4 are the payload to send to/read from the load cell.
* In write mode, MCU writes bytes 2-4 to the load cell.
* In read mode, the load cell sends the requested data as bytes 2-4 to MCU.

The I2C is address is stored bits`[9:3]` in word 02 of the EEPROM.

Step 1: Put sensor into command mode within 6 ms after power up.
1. Write `{0xa0, 0x00, 0x00}` to old I2C address.
2. Check if the transaction is done properly (`if (EnterCommandMode(OldAddress)==true)` ).

Step 2: Request to read word 02 (2 bytes) from the EEPROM.
1. Write (`{0x02, 0x00, 0x00}`) to old I2C address. 
2. Check if the transaction is done properly (`if (ReadEEPROM(OldAddress,0x02)==true)`).

Step 3: Fetch the word requested in step 2.
1. Read from the old I2C address the word 02 into a buffer defined (`EEPROMdata`).
2. Check if transaction is done properly (`if (FetchEEPROM(OldAddress,3, EEPROMdata)==true)`).
3. Verify if the response byte is `0x5a` as expected (`if (buffer[0] == 0x5a)`).
4. The MSB and LSB of the word are stored as `EEPROMdata[1]` and  `EEPROMdata[2]`.

Step 4: Alter the I2C address bits in the word 02 we retrieved.

I think TE's code is a little bit hard to interpret so I'll try to break it down:
1. Assemble the two bytes as one word.
2. Unset bits`[9:3]` (i.e. the 7 bits containing the old address) and bits`[12:10]` (i.e communication lock bits `0b011`) by applying bitmask `0xe007`.
3. Set the new address to bits[9:3].
4.  Set the communication lock bits to bits[12:10].
```c
tempEEPROMdata = EEPROMdata[1] << 8 | EEPROMdata[2];
tempEEPROMdata &= 0xe007; // This unsets bits[12:10]
tempEEPROMdata |= NewAddress << 3;
tempEEPROMdata |= 0xc00;
``` 
Step 5: Write the new word 02 to the load cell. 

1. Break the new word 02 back down to 2 bytes. 
```c
EEPROMdata[1] = (tempEEPROMdata & 0xff00) >> 8; // Apply bitmask 0xff00 and right shift 8 bits.
EEPROMdata[2] = (tempEEPROMdata & 0xff); // Apply bitmask 0xff.
```
2. Write `{0x42, EEPROMdata[1], EEPROMdata[2]}` to the old I2C address.
3. Check if the transaction is done properly (`if ((WriteDataToSensor(OldAddress,3, EEPROMdata)==true)`).

Step 6: Exit command mode.
1. Write `{0x80, 0x00, 0x00} to the old I2C address.
2. Check if the transaction is done properly (`if ((WriteDataToSensor (OldAddress,3, EEPROMdata)==true)`).

**Comment by jakefave** [comment]
You are an incredible individual. Thank you very much. Will give it a go this evening and edit this message with an update tomorrow evening.

**Comment by eepj** [comment]
OK... some bad news.

I tried to change the I2C address of my load cell with the Arduino sketch I mentioned. It seems like the first command must be sent within 6 ms after power up. I wasn't able to manually connect the load cell within that 6 ms time window so the sketch won't run beyond step 3.

I'll try to power the load cell via a load switch and see if I can get the sketch working next week.

**Comment by grunderhunter** [comment]
Hi, firstly thank you for producing this library. I got everything working OK with my load cell using it.

I've got four cells I want to use (no multiplexor) and tried the address change sketch. I get the same issue with the 6ms time window and was wondering if you got it working?

I'm very new to Arduino and I2C sensors, so the only thing I could think that might work is by switching the power to the load cell via a transistor controlled by the Arduino? I'll maybe try it and see, but I think you have a much better chance of getting it working than I do!

**Comment by eepj** [comment]
Hi. I haven't had the time to try that yet, but (TE and) I would recommend against changing the internal registers.

I've received reports (in another issue) that the I2C address of the device was indeed changed by running the sketch but this change could not be reverted.

Using an I2C multiplexer seems to be a safer option for now. Nevertheless I will try to run the sketch via a transistor when I have the time. Thank you for your interest in this library.
